### ISSUE TITLE
The method to generate debug.js breaks git-based updates of Dart-Code versions

### ISSUE THREAD

**Issue by I-am-a-squirrel** [Dart-Code/Dart-Code#4164]

**Comment by DanTup** [comment]
@I-am-a-squirrel can you provide more information about the problem? It's not very clear from the title.

**Comment by I-am-a-squirrel** [comment]
Well, I am using it as a packer.nvim package in Neovim to debug my Flutter application. This package manager works as a wrapper around git protocol. But after git clone I need to generate debug.js file (a debugger file) (https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#Dart) which is generated inside repository. After some updates in the Github repo it causes merge conflicts.

**Comment by DanTup** [comment]
Ah, I see. It's not supported to use Dart-Code's debug adapter outside of this extension, so there are likely some rough edges in trying to do so.

I've been working on moving the debug adapters from Dart-Code into the Dart and Flutter SDKs so they can be more easily reused by other editors. I would recommend having your editor use the SDK-provided debug adapters instead. There are still a few minor issues/gaps between the two, but I suspect they will work for everything you need as long as you're on the latest stable SDKs.

Information can be found here:

- Dart DA: https://github.com/dart-lang/sdk/blob/main/pkg/dds/tool/dap/README.md
- Flutter DA: https://github.com/flutter/flutter/tree/master/packages/flutter_tools/lib/src/debug_adapters#debug-adapter-protocol-dap

I'm not familiar with nvim-dap so can't test it, but I suspect the config you'd want is more something like:

```
  dap.adapters.dart = {
    type = "executable",
    command = "dart", // use "flutter" here for Flutter apps
    args = {"debug_adapter"}
  }
  dap.configurations.dart = {
    {
      type = "dart",
      request = "launch",
      name = "Launch flutter",
      program = "${workspaceFolder}/lib/main.dart",
      cwd = "${workspaceFolder}",
    }
  }
```

There may be issues using this with Flutter < 3.3 so please be sure you're on the latest stable if you're going to use this.

**Comment by MitaiDart** [comment]
> Ah, I see. It's not supported to use Dart-Code's debug adapter outside of this extension, so there are likely some rough edges in trying to do so.
> 
> I've been working on moving the debug adapters from Dart-Code into the Dart and Flutter SDKs so they can be more easily reused by other editors. I would recommend having your editor use the SDK-provided debug adapters instead. There are still a few minor issues/gaps between the two, but I suspect they will work for everything you need as long as you're on the latest stable SDKs.
> 
> Information can be found here:
> 
> * Dart DA: https://github.com/dart-lang/sdk/blob/main/pkg/dds/tool/dap/README.md
> * Flutter DA: https://github.com/flutter/flutter/tree/master/packages/flutter_tools/lib/src/debug_adapters#debug-adapter-protocol-dap
> 
> I'm not familiar with nvim-dap so can't test it, but I suspect the config you'd want is more something like:
> 
> ```
>   dap.adapters.dart = {
>     type = "executable",
>     command = "dart", // use "flutter" here for Flutter apps
>     args = {"debug_adapter"}
>   }
>   dap.configurations.dart = {
>     {
>       type = "dart",
>       request = "launch",
>       name = "Launch flutter",
>       program = "${workspaceFolder}/lib/main.dart",
>       cwd = "${workspaceFolder}",
>     }
>   }
> ```
> 
> There may be issues using this with Flutter < 3.3 so please be sure you're on the latest stable if you're going to use this.

maybe you could help me with the configuration?
https://github.com/mfussenegger/nvim-dap/discussions/785#discussion-4672832

**Comment by DanTup** [comment]
@MitaiDart your configuration (using `flutter debug_adapter`) seems correct to me. It's not clear to me what the problem is though. There are lots of warnings printed, but I don't think those should cause any problems (my guess is that the DAP client is producing warnings for events it doesn't recognise, rather than just ignoring them).

Although, I don't understand why you need to use `flutter run` first. The DAP server allows you to launch applications, not only attach to them. This seems like an unnecessary step.

**Comment by MitaiDart** [comment]
> @MitaiDart your configuration (using `flutter debug_adapter`) seems correct to me. It's not clear to me what the problem is though. There are lots of warnings printed, but I don't think those should cause any problems (my guess is that the DAP client is producing warnings for events it doesn't recognise, rather than just ignoring them).
> 
> Although, I don't understand why you need to use `flutter run` first. The DAP server allows you to launch applications, not only attach to them. This seems like an unnecessary step.

![Снимок экрана от 2022-12-17 22-44-52](https://user-images.githubusercontent.com/104931116/208254647-8011a339-c879-4164-8ff4-149c1388e51f.png)

**Comment by DanTup** [comment]
@MitaiDart have you checked those logs? It would be useful to capture all of the traffic between the client and server to see what's happening. I don't know how to do this in nvim though, so if `dap.set_log_level` doesn't let you log all requests, you might need to ask in their repo how to do this.

**Comment by MitaiDart** [comment]
> dap.configurations.dart

unfortunately, I do not know how, and the Dap community is inactive, there is not a single answer in my discussion, but I found da.luta from a good vim coder, and there is no dap.adapters.dart or dap.configurations.dart settings there
https://github.com/akinsho/dotfiles/blob/d3ec6519dd966220311aa99f0b49aa68b6a4a82e/.config/nvim/lua/as/plugins/dap.lua

**Comment by MitaiDart** [comment]
https://github.com/akinsho/flutter-tools.nvim/blob/main/lua/flutter-tools/dap.lua
it seems to come bundled with flutter-tools

**Comment by DanTup** [comment]
@MitaiDart that's correct, the Dart/Flutter debug adapters ship with the Dart/Flutter SDKs. The config you've linked above looks reasonable to me (it's using `flutter debug-adapter` to start the DAP).

Interestingly it also includes:

```
      options = { -- Dartls is slow to start so avoid warnings from nvim-dap
        initialize_timeout_sec = 30,
      },
```

Perhaps this is the cause of the warning you see, although I'm not sure why initialize would take so long. If someone is able to capture a log of the DAP traffic where this is slow, I would be interested in investigating to see if it can be fixed. @akinsho I see you increased this from 6 to 30s at one point. Can you confirm whether this specifically refers to the `initialize` request, or could it also cover responding to the `launch` request? (if the latter, this may be fixed by https://github.com/flutter/flutter/pull/109386/files but it's not shipped in a stable Flutter SDK yet).

**Comment by MitaiDart** [comment]
I was able to fasten it, but there are no windows with information... and so everything works)))
![Снимок экрана от 2022-12-18 20-37-55](https://user-images.githubusercontent.com/104931116/208307073-5563fbef-b628-4003-a85e-3a80f3563f1b.png)

**Comment by DanTup** [comment]
@MitaiDart do you mean everything is working now, and the issue is resolved?

**Comment by MitaiDart** [comment]
> @MitaiDart do you mean everything is working now, and the issue is resolved?

in general, yes, it still gives out a bunch of errors and warnings there, but it seems to work))

**Comment by akinsho** [comment]
@DanTup sorry for the very late response, regarding the question you asked
> I see you increased this from 6 to 30s at one point. Can you confirm whether this specifically refers to the initialize request, or could it also cover responding to the launch request? (if the latter, this may be fixed by [flutter/flutter#109386 (files)](https://github.com/flutter/flutter/pull/109386/files) but it's not shipped in a stable Flutter SDK yet).

The timeout only refers to the `initialize` request, I found bumping in in my plugin didn't really make a huge difference but also the error I believe is harmless because everything still works as expected and I think could have been the debug plugin being a little ambitious with the expected startup times, although I don't really know.

As a side note thanks for your work to split out the debugger into the flutter/dart SDKs that has been hugely beneficial for the (n)vim community 🙏🏿

**Comment by DanTup** [comment]
@akinsho np! Since that last comment, I expect https://github.com/flutter/flutter/pull/109386 made it to stable, so the `launch` request shouldn't take so long to respond now.

If you run into any other issues using the debug adapters, please do let me know (either filing issues here, or in dart-lang/sdk, and feel free to CC me in them..).
